eip-0024.md

```diff
@@ -7,24 +7,125 @@
+
+- R4 of this box is an Integer, showing 1000 * royalty percentage of the artwork. e.g, 20 for 2% royalty. If R4 of the issuer box of an artowrk is empty or non-positive, then royalty is considered to be 0%.
+- In the current design of the [Auction contranct](eip-0022.md), proposition bytes of this box is the contract that the royalty percentage will be sent to. This is a suboptimal design because of AOT costing mechanism which is supposed to be replaced with JIT costing with a soft-fork that will happen with v5.0 of [ergo node](https://github.com/ergoplatform/ergo).
+  
+As mentioned, royalty amount will go to the propositiona bytes of the issuer box. In the case of a simple proxy contract (proposition bytes of the issuer box), this means that the artist will receive the royalty share. However, the proxy contract may enforce the royalty to go to any other complex contract - e.g., 20% to the artist, and 80% to some charity.
+
+## Design V2
+Version 2 of this proposal is created to both extend this proposal to include collections and fix some poor design choices that were caused by the AOT costing issues. The following is the v2 standard for this box.
+- R4 of this box is an Integer, showing 1000 * royalty percentage of the artwork. e.g, 20 for 2% royalty. If R4 of the issuer box of an artwork is empty or non-positive, then royalty is considered to be 0%.
+- R5 of this box is an integer showing the artwork standard version. If This register is empty, or the type is something other than integer, then it should be considered as part of V1 design. Otherwise, this register shows the version number for the standard.
+- R6 of this box is of type Coll[(Coll[Byte], Int)] that shows information about royalty, in particular, a list of royalty receivers and their respective share. The `Coll[Byte]` part is the ergo tree of one of the royalty receivers and the `Int` part is a number between 1 and 100 showing the share of this receiver. The sum of shares for royalty receivers must add up to 100.
+In this version of the standard, unlike V1, we don't specify how the royalty will be sent to the receivers. In V1, the proposition bytes of the issuer box receive the royalty amount. However, in this version, the royalty amounts may go to receivers specified in R6 or the proposition bytes of the issuer box depending on the [auction contract](eip-0022.md) standard. So NFT developers need to be careful about both R6 and the proposition bytes of the issuer box to be compatible.
+- R7 of this box may contain a JSON specifying trait values for the NFT as Coll[Byte]. 
+There are three types of traits:
+  - Properties: These are textual traits such as specifying `sex` as `male`.
+  - Levels: These are numerical traits that encode some sort of level-like information such as specifying `speed` as `60 out of 100`.
+  - Stats: These are numerical traits that encode any numerical information about the NFT such as specifying `age` as `25 out of 50`.
```

Comment:
Using Json have two major drawbacks. 
1) it is not compact, a lot of space is wasted by storing repeated "key", "value", "max" names, which are not really necessary.
2) The content cannot be accessed from smart contracts (this may be not necessary now, but no doubt will find it's use case in future)

I suggest to use the following Sigma type, which is isomorphic to the json below.

```
( // properties
  Coll[(  
    Coll[Byte],   // key
    Coll[Byte]    // value
  )],
  (  // levels
     Coll[(
        Coll[Byte], // key
        (Int, Int)     // value, max
      )],
     // stats
     Coll[(
        Coll[Byte], // key
        (Int, Int)     // value, max
      )]
  )
)
```

A library of generic encoders/decoders can be implemented in Appkit (and in JS if necessary) to implement such canonical conversions Json <-> ErgoValue and avoid doing it in all dApps again and again.

Reply:
This is a great idea!

Reply:
@aslesarenko It's actually a great idea.

Reply:
@aslesarenko how would Json <-> ErgoValue work for converting ErgoValue to Json if we have no keys in ErgoValue?

Reply:
@anon-real If you are talking about names of Json fields, then they are should be known from the context (basically they are static metadata, in my example above I put the names in comments).
If something is not static, then it is data and should be stored in the data structure itself.